scalalib/src/mill/scalalib/CoursierModule.scala

```diff
@@ -21,8 +21,12 @@ trait CoursierModule extends mill.Module {
    * Bind a dependency ([[Dep]]) to the actual module contetxt (e.g. the scala version and the platform suffix)
    * @return The [[BoundDep]]
    */
-  def bindDependency: Task[Dep => BoundDep] = T.task { dep: Dep =>
-    BoundDep((resolveCoursierDependency(): @nowarn).apply(dep), dep.force)
+  def bindDependency(platformed: Boolean): Task[Dep => BoundDep] = T.task { dep: Dep =>
```

Comment:
Can you motivate the extra `platformed` flag? I was hoping, we can avoid to decide with every use of `bindDependency` whether we need to enable platformed. What are the use cases?

Reply:
@lefou the basic issue is that someone somewhere needs to decide which dependencies within e.g. a `ScalaJSModule` need to be platform-bound, and which don't:

1. Scaladoc, Scala-compiler, Scala-compiler-plugins *should not* be platform bound

2. `compileIvyDeps`, `runIvyDeps` *should* be platform bound

Approaches:

1. The status quo in master is that this distinction is done on a dep-by-dep basis, via `ivy"foo::bar:baz"` vs `ivy"foo::bar::baz"`. This has the downside it is confusing, and allows you to mix platformed/non-platformed dependencies in the same resolved classpath which is basically never what you want

2. My earlier update to this PR preserved that approach, except making `ivy"foo::bar:baz"` do what `ivy"foo::bar::baz"` does  and adding `ivy"foo::bar:baz".withPlatformed(false)` to do what  `ivy"foo::bar:baz"`  used to do. This is less confusing, but still has the weakness of being able to mix platformed and non-platformed dependencies

3. This update in response to your feedback moves the `withPlatformed` logic inside `bindDependency`, but we still need to tell `bindDependency` somehow whether these are platform-specific dependencies or not. Hence the extra flag. Since `bindDependency` is typically called once per classpath (whether compile, run, compiler, scaladoc, etc.), this has the added advantage that it becomes impossible to accidentally mix platformed and non-platformed dependencies

But the bottom line is we need to specify which dependencies are platformed and which are not, which is both necessary and not something that can be inferred from the other data we already have

Reply:
Another place we could potentially put the distinction is when we call `resolveDeps`, but AFAICT `resolveDeps` and `bindDependency` are ~always called together anyway, so maybe it's a moot point

Reply:
Yeah, I think because there are various differing use site, we initially ended up with the explicit demarcation, which I still think is the best approach. Only the choosen `::` is a bit too insignificant and not easy to spot. 

About the distinction of `bindDependency` and `resolveDeps`, both have distinct use cases and should not be merged.

`bindDependency` is needed to bind a rather loose dependency declaration to the current module, such that it can be used in transitive contexts, e.g. where no Scala version is defined. Hence, its call scope is per-module in contrast to `resolveDeps`, which is called per-classpath.

`resolveDeps` is machinery to resolve and fetch some already bound dependencies (which were defined in potentially different modules). (Eventually but unrelated, we may want to split this even further into resolve and fetch, so we can better simulate lock file semantics and improve cache granularity.)

I think we should keep the platform-decision in the `Dep` class. If we need a third state (set, unset, don't care) we can model it there, e.g. with an `Option[Boolean]`. I think the `bindDependency` task should only depend on other tasks in the same module, not a flag.

Reply:
The best place to default-bind the `platform` flag are probably the various `resolvedIvyDeps`, `resolvedRunIvyDeps`, `resolvedScalacPluginIvyDeps` tasks, which all call the `resolveDeps` task. We may need to introduce a dedicated `resolvedScaladocIvyDeps` to honor the fact that these may always need non-platformed dependencies (in the JS, Native cases, at least).

Reply:
@lefou going through the code, it definitely seems like `bindDependency` is the right place to do this. After all, `bindDependency`'s doc says

> Bind a dependency ([[Dep]]) to the actual module contetxt (e.g. the scala version and the platform suffix)
>   * @return The [[BoundDep]]

With `BoundDep`

> Same as [[Dep]] but with already bound cross and platform settings.

Because `BoundDep` already has all artifact name suffixes already "flattened out" and appended to the artifact name, it is impossible to robustly take an existing `BoundDep` and add the platform suffix: we would have to e.g. take `scala-async_2.13`, split it into `scala-async` and `2.13`, and then splice in `scala-sjs1` in the middle to make it `scala-async_sjs1_2.13`. Maybe possible to do hackily for the suffices we use now, but would be very fragile e.g. if someone's platform suffix or artifact name has a `_` everything will break.

As is, the current state of this PR already basically performs the `platformed` decision everywhere close to we call `resolveDeps`, It's just that the usage of `resolveDeps` is a bit more heterogenous than the small number of tasks you listed above. We can refactor the call graph later if we want, but fundamentally it does seem that performing the `platformed=true` or `platformed=false` decision does belong inside `bindDependency` and nowhere else